src/backend/distributed/transaction/worker_transaction.c

```diff
@@ -122,437 +82,113 @@ SendCommandToWorkersParams(TargetWorkerSet targetWorkerSet, char *command,
+		if (targetWorkerSet == WORKERS_WITH_METADATA && !workerNode->hasMetadata)
 		{
-			ReraiseRemoteError(workerConnection, queryResult);
+			continue;
 		}
 
-		PQclear(queryResult);
-
-		/* clear NULL result */
-		PQgetResult(workerConnection);
+		connection = StartNodeUserDatabaseConnection(connectionFlags, nodeName, nodePort,
+													 nodeUser, NULL);
 
-		/* we no longer need this connection */
-		CloseConnectionByPGconn(workerConnection);
-	}
-	PG_CATCH();
-	{
-		/* close the connection */
-		CloseConnectionByPGconn(workerConnection);
+		MarkRemoteTransactionCritical(connection);
 
-		PG_RE_THROW();
+		connectionList = lappend(connectionList, connection);
 	}
-	PG_END_TRY();
-}
-
 
-/*
- * IsWorkerTransactionActive returns true if there exists any on going
- * worker transactions.
- */
-bool
-IsWorkerTransactionActive(void)
-{
-	bool isWorkerTransactionActive = false;
-
-	if (workerConnectionList != NIL)
-	{
-		isWorkerTransactionActive = true;
-	}
-
-	return isWorkerTransactionActive;
-}
-
-
-/*
- * RemoveWorkerTransaction removes the transaction connection to the specified node from
- * the transaction connection list.
- */
-void
-RemoveWorkerTransaction(char *nodeName, int32 nodePort)
-{
-	TransactionConnection *transactionConnection =
-		GetWorkerTransaction(nodeName, nodePort);
-
-	/* transactionConnection = NULL if the worker transactions have not opened before */
-	if (transactionConnection != NULL)
+	/* finish opening connections */
+	foreach(connectionCell, connectionList)
 	{
-		PGconn *connection = transactionConnection->connection;
-
-		/* closing the connection will rollback all uncommited transactions */
-		CloseConnectionByPGconn(connection);
+		MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
 
-		workerConnectionList = list_delete(workerConnectionList, transactionConnection);
+		FinishConnectionEstablishment(connection);
 	}
-}
 
+	RemoteTransactionsBeginIfNecessary(connectionList);
 
-/*
- * EnableXactCallback registers the CompleteWorkerTransactions function as the callback
- * of the worker transactions.
- */
-static void
-EnableXactCallback(void)
-{
-	if (!isXactCallbackRegistered)
+	/* send commands in parallel */
+	foreach(connectionCell, connectionList)
 	{
-		RegisterXactCallback(CompleteWorkerTransactions, NULL);
-		isXactCallbackRegistered = true;
-	}
-}
-
+		MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
 
-/*
- * CompleteWorkerTransaction commits or aborts pending worker transactions
- * when the local transaction commits or aborts.
- */
-static void
-CompleteWorkerTransactions(XactEvent event, void *arg)
-{
-	if (workerConnectionList == NIL)
-	{
-		/* nothing to do */
-		return;
-	}
-	else if (event == XACT_EVENT_PRE_COMMIT)
-	{
-		if (MultiShardCommitProtocol == COMMIT_PROTOCOL_2PC)
+		int querySent = PQsendQueryParams(connection->pgConn, command, parameterCount,
+										  parameterTypes, parameterValues, NULL, NULL, 0);
+		if (querySent == 0)
```

Comment:
Doesn't have to be now, but we really should have a wrapper for PQsendQueryParams too, marking the connection as nonblocking before sending the command. Otherwise we can stall here too :(